pinot-segment-local/src/main/java/org/apache/pinot/segment/local/segment/index/loader/ForwardIndexHandler.java

```diff
@@ -0,0 +1,269 @@
+      // Remove inverted index if exists.
+      // For v1 and v2, it's the actual inverted index. For v3, it's the temporary inverted index.
+      FileUtils.deleteQuietly(fwdIndexFile);
+    }
+
+    LOGGER.info("Creating new forward index for segment={} and column={}", segmentName, column);
+
+    Map<String, ChunkCompressionType> compressionConfigs = _indexLoadingConfig.getCompressionConfigs();
+    Preconditions.checkState(compressionConfigs.containsKey(column));
+    // At this point, compressionConfigs is guaranteed to contain the column.
+    ChunkCompressionType newCompressionType = compressionConfigs.get(column);
+
+    int numDocs = existingColMetadata.getTotalDocs();
+
+    try (ForwardIndexReader reader = LoaderUtils.getForwardIndexReader(segmentWriter, existingColMetadata)) {
+      int lengthOfLongestEntry = reader.getLengthOfLongestEntry();
+      Preconditions.checkState(lengthOfLongestEntry >= 0,
+          "lengthOfLongestEntry cannot be negative. segment=" + segmentName + " column={}" + column);
+
+      IndexCreationContext.Forward context =
+          IndexCreationContext.builder().withIndexDir(indexDir).withColumnMetadata(existingColMetadata)
+              .withLengthOfLongestEntry(lengthOfLongestEntry).build()
+              .forForwardIndex(newCompressionType, _indexLoadingConfig.getColumnProperties());
+
+      try (ForwardIndexCreator creator = indexCreatorProvider.newForwardIndexCreator(context)) {
+        PinotSegmentColumnReader columnReader =
+            new PinotSegmentColumnReader(reader, null, null, existingColMetadata.getMaxNumberOfMultiValues());
+
+        for (int i = 0; i < numDocs; i++) {
+          Object val = columnReader.getValue(i);
+
+          // JSON fields are either stored as string or bytes. No special handling is needed.
+          switch (creator.getValueType()) {
+            case INT:
+              creator.putInt((int) val);
+              break;
+            case LONG:
+              creator.putLong((long) val);
+              break;
+            case FLOAT:
+              creator.putFloat((float) val);
+              break;
+            case DOUBLE:
+              creator.putDouble((double) val);
+              break;
+            case STRING:
+              creator.putString((String) val);
+              break;
+            case BYTES:
+              creator.putBytes((byte[]) val);
+              break;
+            case BIG_DECIMAL:
+              creator.putBigDecimal((BigDecimal) val);
+              break;
+            default:
+              throw new IllegalStateException();
+          }
+        }
+      }
+    }
+
+    // We used the existing forward index to generate a new forward index. Existing forward index is not needed
+    // anymore. Remove it.
+    // Note that the stale entries corresponding to old forward index in columns.psf file will be removed when
+    // segmentWriter.close() is called.
+    segmentWriter.removeIndex(column, ColumnIndexType.FORWARD_INDEX);
```

Comment:
I am a bit confused by this comment of "removing stale entries when `segmentWriter.close` is called". What cleanup are we relying on `segmentWriter` for ?

Reply:
Clarified the comment. Please let me know if this makes sense.